src/NumericalAlgorithms/LinearSolver/Gmres.hpp

```diff
@@ -0,0 +1,209 @@
+                      const SourceType& source, const VarsType& initial_guess,
+                      const Preconditioner& preconditioner =
+                          IdentityPreconditioner<VarsType>{}) const noexcept {
+    constexpr bool use_preconditioner =
+        not cpp17::is_same_v<Preconditioner, IdentityPreconditioner<VarsType>>;
+
+    auto result = initial_guess;
+    convergence_reason_ = boost::none;
+    iteration_ = 0;
+    // Can allocated memory for the operand_ be kept around for successive
+    // solves?
+    VarsType operand_{};
+    while (not convergence_reason_) {
+      operand_ = source - linear_operator(result);
+      const double initial_residual_magnitude =
+          sqrt(inner_product(operand_, operand_));
+      convergence_reason_ = Convergence::criteria_match(
+          convergence_criteria_, iteration_, initial_residual_magnitude,
+          initial_residual_magnitude);
+      if (convergence_reason_) {
+        break;
+      }
+      operand_ /= initial_residual_magnitude;
+      basis_history_[0] = operand_;
+
+      std::pair<DenseVector<double>, double> minres_and_magnitude{};
+      for (size_t k = 0; k < restart_; k++) {
+        if (use_preconditioner) {
+          preconditioned_basis_history_[k] = preconditioner(operand_);
+        }
+        operand_ = linear_operator(
+            use_preconditioner ? preconditioned_basis_history_[k] : operand_);
+        // Resize matrix
+        orthogonalization_history_.resize(k + 2, k + 1);
+        // Make sure the new entries are zero
+        for (size_t i = 0; i < orthogonalization_history_.rows(); i++) {
+          orthogonalization_history_(
+              i, orthogonalization_history_.columns() - 1) = 0.;
+        }
+        for (size_t j = 0; j < orthogonalization_history_.columns(); j++) {
+          orthogonalization_history_(orthogonalization_history_.rows() - 1, j) =
+              0.;
+        }
+        // Arnoldi orthogonalization
+        for (size_t j = 0; j < k + 1; j++) {
+          const double orthogonalization =
+              inner_product(basis_history_[j], operand_);
+          orthogonalization_history_(j, k) = orthogonalization;
+          operand_ -= orthogonalization * basis_history_[j];
+        }
+        orthogonalization_history_(k + 1, k) =
+            sqrt(inner_product(operand_, operand_));
+        // Least-squares solve for the minimal residual
+        minres_and_magnitude = gmres_detail::minimal_residual(
+            orthogonalization_history_, initial_residual_magnitude);
+        iteration_++;
+        convergence_reason_ = Convergence::criteria_match(
+            convergence_criteria_, iteration_, minres_and_magnitude.second,
+            initial_residual_magnitude);
+        if (UNLIKELY(convergence_reason_)) {
+          break;
+        } else if (k + 1 < restart_) {
+          operand_ /= orthogonalization_history_(k + 1, k);
+          basis_history_[k + 1] = operand_;
```

Comment:
at least the assignment of this could be moved to the beginning of the loop, removing the pre-loop assignment as well. Arguably, you could do the same for the `/=` by just reusing an entry of the `orthogonalization_history_` for the `initial_residual_magnitude` normalization at the start, but I'll leave the second one up to you because it potentially undermines the conceptual distinction between the objects.